Exam/CT/ds/src/exercice3/view/JBox.java

```diff
@@ -0,0 +1,121 @@
+package exercice3.view;
+
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.Image;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+import java.awt.image.BufferedImage;
+import java.io.IOException;
+import java.util.Observable;
+import java.util.Observer;
+
+import javax.imageio.ImageIO;
+import javax.swing.BorderFactory;
+import javax.swing.ImageIcon;
+import javax.swing.JButton;
+import javax.swing.border.BevelBorder;
+import javax.swing.plaf.metal.MetalButtonUI;
+
+import exercice2.Util;
+import exercice3.model.Box;
+import exercice3.model.BoxState;
+
+public class JBox extends JButton implements Observer {
+
+	private static final Dimension PREFERED_SIZE = new Dimension(24, 24);
+	private static ImageIcon MINE,FLAG;
+
+	private final Box model;
+
+	public JBox(Box model) {
+		this.model = model;
+		this.model.addObserver(this);
+
+		setMinimumSize(PREFERED_SIZE);
+		setPreferredSize(PREFERED_SIZE);
+
+		addMouseListener(new MouseAdapter() {
+			public void mouseClicked(MouseEvent e) {
+				if (model.getState() == BoxState.TRIGGERED) {
+					if (e.getButton() == MouseEvent.BUTTON3) {
+						// TODO: Implement ASAP
+						model.flag();
+						imgFlag();
+					} else {
+					return;	
+					}
+				}
+				else if (e.getButton() == MouseEvent.BUTTON1) {
+					model.trigger();
+				} else if (e.getButton() == MouseEvent.BUTTON3) {
+					// TODO: Implement ASAP
+					model.flag();
```

Comment:
This implementation unfortunately does not allow to handle properly all the cases...
And you have some weird side effects like: box *looks* clicked when you set a flag.